src/geometry_base/kd_tree_strategy.cc

```diff
@@ -0,0 +1,143 @@
+//   this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "maliput/geometry_base/kd_tree_strategy.h"
+
+#include <cstdlib>
+
+#include "maliput/math/kd_tree.h"
+
+namespace maliput {
+namespace geometry_base {
+
+api::RoadPositionResult KDTreeStrategy::DoToRoadPosition(const api::InertialPosition& inertial_position,
+                                                         const std::optional<api::RoadPosition>& hint) const {
+  if (hint.has_value()) {
+    MALIPUT_THROW_UNLESS(hint->lane != nullptr);
+    const maliput::api::LanePositionResult lane_pos = hint->lane->ToLanePosition(inertial_position);
+    return {{hint->lane, lane_pos.lane_position}, lane_pos.nearest_position, lane_pos.distance};
+  }
+  const auto lane_id = this->ClosestLane(inertial_position.xyz());
+  const auto lane = rg_->ById().GetLane(lane_id);
+  const auto lane_position_result = lane->ToLanePosition(inertial_position);
+  return {
+      {lane, lane_position_result.lane_position}, lane_position_result.nearest_position, lane_position_result.distance};
+}
+
+std::vector<api::RoadPositionResult> KDTreeStrategy::DoFindRoadPositions(const api::InertialPosition& inertial_position,
+                                                                         double radius) const {
+  const auto lane_ids = this->ClosestLanes(inertial_position.xyz(), radius);
+  std::vector<api::RoadPositionResult> road_positions;
+  for (const auto& current_lane : lane_ids) {
+    const auto lane = rg_->ById().GetLane(current_lane);
+    const auto lane_position = lane->ToLanePosition(inertial_position);
+    if (lane_position.distance <= radius) {
+      api::RoadPositionResult road_position{
+          {lane, lane_position.lane_position}, lane_position.nearest_position, lane_position.distance};
+      road_positions.push_back(road_position);
+    }
+  }
+  return road_positions;
+}
+
+void KDTreeStrategy::Init() {
+  const auto lanes = rg_->ById().GetLanes();
+
+  std::deque<maliput::geometry_base::MaliputPoint> points;
+  for (const auto& lane : lanes) {
+    const auto lane_length = lane.second->length();
+    for (double s = 0; s <= lane_length; s += 0.1) {
```

Comment:
Steps in `s` should be set in terms of linear_tolerance. You could use twice linear_tolerance.
Alternatively, you could opt to pass in that value at  construct time and let it as a configuration parameter. I would prefer that because this constant here is a) arbitrary and b) may dramatically impact the overall construction time.

Reply:
Regarding the option of passing the value as a configuration parameter during construction time, I don't think it is possible because it is not something that all the strategies need so we would enter again the realm of redesigning in order to do it that way.

Regarding the use of linear_tolerance to define the step size instead of the arbitrary 10cm, that should be possible, but, I'm worried about construction time, since it can greatly increase depending on the selected linear_tolerance. I'll do a couple of local test modifying the value and using the linear_tolerance and see how it goes in comparison.

Reply:
See https://en.cppreference.com/w/cpp/memory/unique_ptr/make_unique and how it is defined for example.

We could define a similar function in maliput::geometry_base::RoadGeometry to build strategies and forward arguments to the constructors. The only condition is that the template arguemnt is a base of Strategy and that can be solved at compile time with `static_assert()` and `std::is_base_of()` --> https://en.cppreference.com/w/cpp/types/is_base_of

Reply:
> Regarding the use of linear_tolerance to define the step size instead of the arbitrary 10cm, that should be possible, but, I'm worried about construction time, since it can greatly increase depending on the selected linear_tolerance. I'll do a couple of local test modifying the value and using the linear_tolerance and see how it goes in comparison.

With the previous approach, we can set the sampling step, and dynamic sampling policies as we need to make more complex strategies.

Reply:
You are right, we can solve it with a variadic function template. Tho I am having doubts about passing that step value in runtime. I guess that it would be ok as the type will be defined in compilation time, right?

Reply:
If I understood correctly we would leave the selection of the sampling step to the builder of the road geometry. Take a look at the implementation. Done.